### PR TITLE
Remove rAF, also support for uncontrolled mode and valueLink

### PR DESCRIPTION
**Pull request by andreypopp**
Not sure why we needed rAF there.

Also I think it's better to remove support for unconrolled mode and valueLink. (I think this makes code much simpler and easier to reason about). Thoughts?

### CODE REVIEW

File: src/calculateNodeHeight.js [hunk 1]
```diff
@@ -35,9 +35,12 @@ const SIZING_STYLE = [
 let computedStyleCache = {};
 let hiddenTextarea;
 
-export default function calculateNodeHeight(uiTextNode,
+export default function calculateNodeHeight(
+    uiTextNode,
+    value,
     useCache = false,
     minRows = null, maxRows = null) {
+  console.log('calculateNodeHeight', JSON.stringify(value));
```
Comment (insin) [inline src/calculateNodeHeight.js]: Leftover debug logging?

Comment (andreypopp) [inline src/calculateNodeHeight.js]: Thanks

**Comment by andreypopp** [comment]
Hm... actually we might need to add rAF back to batch `resize` event?

**Comment by julen** [comment]
I'm using autosize as an uncontrolled component because I'm altering the value being input and I need to hack around the cursor position (see https://github.com/facebook/react/issues/955#issuecomment-160831548 for reference). Would this PR mean I cannot use the component in such a context anymore? That'd be a bummer 😕

**Comment by Keats** [comment]
Any update on that?

**Comment by slorber** [comment]
@andreypopp with release 4.0.2 and React 14, I had the Textarea to render twice, leading to the cursor going at the end of the input. This is because `_resizeComponent` was called twice (during both `_onChange` and `componentWillReceiveProps`). rAF may be a problem too in `componentWillReceiveProps`




Do you plan to release a fix for that soon?

**Comment by slorber** [comment]
As a suggestion I think Om wraps all his inputs/textarea inside a custom component and avoid render them if value has not changed. This way it's not a big deal if you call setState too much on the parent or use rAF, leading to async rerenders of the input

**Comment by slorber** [comment]
@andreypopp for the record, I just published the fork based on this branch: `react-textarea-autosize-fork-slorber@4.0.1`

This branch fixed my cursor problem because you avoid calling setState when only the input state changes:

```javascript
  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) { // before it always trigger the resize
      this._resizeComponent(nextProps.value || nextProps.placeholder || '');
    }
  }
```

Please ping me when this is released so that I can stop using that fork and switch back to the original project :)

**Comment by Keats** [comment]
@andreypopp need any help for that?

**Comment by slorber** [comment]
@Keats help has been asked here: https://github.com/andreypopp/react-textarea-autosize/issues/115

**Pull request closed without merging**